crates/bevy_ecs/src/query/iter.rs

```diff
@@ -1313,25 +1325,29 @@ impl<'w, 's, D: QueryData, F: QueryFilter, I: Iterator<Item: Borrow<Entity>>>
                 continue;
             };
 
-            if !query_state
-                .matched_archetypes
-                .contains(location.archetype_id.index())
-            {
-                continue;
-            }
+            if *previous_archetype_id != location.archetype_id {
+                if !query_state
```

Comment:
shouldn't this check be outside the new if statement?

Reply:
We only cache valid archetype ids, so if an entity's archetype id matches the last, it must match the query.

However this depends on us initializing the cache with an archetype id that cannot be returned by any possible input entity.

I was under the impression that this held for `ArchetypeId::INVALID`, but I realize now that this id can be returned by pending entities.

Does `ArchetypeId::EMPTY` work for initialization instead?

Reply:
Actually, it seems that the note on `Entities::get` about pending entities is incorrect? Which means no need to change anything here.
I hope nothing depends on that noted behavior...

Reply:
That makes sense. It'd be nice if we could test this, but I'm not sure if it's even possible to create an entity in the invalid archetype.

Reply:
Hmm, you might be able too between a despawn and a flush.

Reply:
To be clear, this is the implementation of `Entities::get`:
```rust
/// Returns the location of an [`Entity`].
    /// Note: for pending entities, returns `Some(EntityLocation::INVALID)`.
    #[inline]
    pub fn get(&self, entity: Entity) -> Option<EntityLocation> {
        if let Some(meta) = self.meta.get(entity.index() as usize) {
            if meta.generation != entity.generation
                || meta.location.archetype_id == ArchetypeId::INVALID
            {
                return None;
            }
            Some(meta.location)
        } else {
            None
        }
    }
```
Any invalid `ArchetypeId` is filtered without exception, the note might have been referring to the validity of the other `EntityLocation` fields.

Reply:
If we can call `iter_many` while an archetype (or other location fields) is invalid, we'll have to remove some `debug_checked_unwrap`s